lib/Sema/CSSimplify.cpp

```diff
@@ -4768,6 +4800,145 @@ ConstraintSystem::simplifyKeyPathApplicationConstraint(
   return unsolved();
 }
 
+Type ConstraintSystem::simplifyAppliedOverloads(
+                                Type fnType,
+                                const FunctionType *argFnType,
+                                Optional<ArgumentLabelState> argumentLabels,
+                                ConstraintLocatorBuilder locator) {
+  auto fnTypeVar = fnType->getAs<TypeVariableType>();
+  if (!fnTypeVar)
+    return fnType;
+
+  // Always work on the representation.
+  fnTypeVar = getRepresentative(fnTypeVar);
+
+  // Dig out the disjunction that describes this overload.
+  auto disjunction = getUnboundBindOverloadDisjunction(fnTypeVar);
+  if (!disjunction) return fnType;
+
+  /// The common result type amongst all function overloads.
+  Optional<Type> commonResultType;
+  auto updateCommonResultType = [&](Type choiceType) {
+    auto markFailure = [&] {
+      commonResultType = Type();
+    };
+
+    auto choiceFnType = choiceType->getAs<FunctionType>();
+    if (!choiceFnType)
+      return markFailure();
+
+    // For now, don't attempt to establish a common result type when there
+    // are type parameters.
+    Type choiceResultType = choiceFnType->getResult();
+    if (choiceResultType->hasTypeParameter())
+      return markFailure();
+
+    // If we haven't seen a common result type yet, record what we found.
+    if (!commonResultType) {
+      commonResultType = choiceResultType;
+      return;
+    }
+
+    // If we already failed, we're done.
+    if (commonResultType->isNull())
+      return;
+
+    // If we found something different, fail.
+    if (!commonResultType.getValue()->isEqual(choiceResultType))
+      return markFailure();
+  };
+
+  // Consider each of the constraints in the disjunction.
+retry_after_fail:
+  bool skippedAnyConstraints = false;
+  bool labelMismatch = false;
+  auto filterResult =
+      filterDisjunctions(disjunction, /*restoreOnFail=*/shouldAttemptFixes(),
+                         [&](Constraint *constraint) {
+        // We must have bind-overload constraints.
+        // FIXME: This isn't entirely true.
+        if (constraint->getKind() != ConstraintKind::BindOverload) {
+          skippedAnyConstraints = true;
+          return true;
+        }
+
+        // We must be binding the type variable (or a type variable
+        // equivalent to it).
+        auto boundTypeVar = constraint->getFirstType()->getAs<TypeVariableType>();
+        if (!boundTypeVar || getRepresentative(boundTypeVar) != fnTypeVar) {
```

Comment:
Should both of these checks be assertions since `getUnboundBindOverloadDisjunction` is supposed to return only viable `bind` disjunctions associated with current type variable?

Reply:
`getUnboundBindOverloadDisjunction()` isn't quite doing all of that checking, but it looks like other clients expect this, too. I'll update it accordingly and simplify `filterDisjunction()`.